### PR TITLE
Better Static Allocation

### PR DESCRIPTION
**Pull request by konbraphat51**
## What changed
#43 
* Visual quality improved
* Speed up by parallelizing by `joblib`
* Add verbosity

### CODE REVIEW

**Comment by konbraphat51** [comment]
current state
sasa_20231227_090830

**Comment by konbraphat51** [comment]
Current result

**Comment by konbraphat51** [comment]
Review please

**Comment by konbraphat51** [comment]
MagneticAllocation become x720 faster (12m -> 3s)

**Comment by konbraphat51** [comment]
current `dev` merged.
Revirew please

**Comment by konbraphat51** [comment]
@SuperHotDogCat

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationCalculator.py [hunk 1]
```diff
@@ -65,7 +65,10 @@ def allocate(
...
     if config.allocation_strategy == "magnetic":
         allocator = MagneticAllocation(
-            config.image_width, config.image_height, config.image_division
+            config.image_width,
+            config.image_height,
+            config.image_division,
+            config.verbosity,
         )
         return allocator.allocate(words, allocation_before)
     else:
```
Comment (SuperHotDogCat) [inline AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationCalculator.py:74]: function estimate_text_size's output type is weird. (int, int)->tuple[int, int]?

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py [hunk 2]
```diff
@@ -281,21 +303,26 @@ def _try_put_all_candidates(
...
-
-            if score > best_score:
-                # best score updated
-                best_position = center_position
-                best_score = score
+        # guard
+        if best_position is None:
+            raise Exception("No position found")
 
         return best_position
```
Comment (SuperHotDogCat) [inline AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py:328]: function _is_hitting_other_words' s size argument is weird. [int, int]->tuple[int, int]?

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py [hunk 3]
```diff
@@ -77,29 +86,34 @@ def allocate(
...
+
         # from second word
-        for word in self.words[1:]:
-            (
-                magnet_outer_frontier,
-                self.rects_outermost,
-            ) = get_magnet_outer_frontier(
-                self.rects_outermost,
+        for word in iterator:
+            magnet_outer_frontier = get_magnet_outer_frontier(
```
Comment (SuperHotDogCat) [inline AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py:109]: Please explain what get_magnet_outer_frontier is called for and what magnet_outer_frontier is.

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationCalculator.py (hunk 1 above)
Comment (konbraphat51) [reply AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationCalculator.py:74]: damm, fixed

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py (hunk 2 above)
Comment (konbraphat51) [reply AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py:328]: fixed

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py (hunk 3 above)
Comment (konbraphat51) [reply AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationCalculator/StaticAllocationStrategies/MagneticAllocation/MagneticAllocation.py:111]: Got it,
Check again please

**Comment by konbraphat51** [comment]
All handled review again please

**Review by SuperHotDogCat — APPROVED** [review]
LGTM

**Pull request merged**